helix-view/src/editor.rs

```diff
@@ -398,18 +409,64 @@ impl Editor {
             .selections
             .remove(&id);
 
-        if close_buffer {
-            // get around borrowck issues
-            let doc = &self.documents[&view.doc];
+        self.tree.remove(id);
+        self._refresh();
+    }
 
-            if let Some(language_server) = doc.language_server() {
-                tokio::spawn(language_server.text_document_did_close(doc.identifier()));
-            }
-            self.documents.remove(&view.doc);
+    pub fn close_document(&mut self, doc_id: DocumentId, force: bool) -> anyhow::Result<()> {
+        let doc = match self.documents.get(&doc_id) {
+            Some(doc) => doc,
+            None => anyhow::bail!("document does not exist"),
+        };
+
+        if !force && doc.is_modified() {
+            anyhow::bail!(
+                "buffer {:?} is modified",
+                doc.relative_path()
+                    .map(|path| path.to_string_lossy().to_string())
+                    .unwrap_or_else(|| "[scratch]".into())
+            );
+        }
+
+        if let Some(language_server) = doc.language_server() {
+            tokio::spawn(language_server.text_document_did_close(doc.identifier()));
+        }
+
+        let views_to_close = self
+            .tree
+            .views()
+            .filter_map(|(view, focus)| {
+                if view.doc == doc_id {
+                    Some((view.id, focus))
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
+
+        for (view_id, _focus) in views_to_close {
+            self.close(view_id);
+        }
+
+        self.documents.remove(&doc_id);
+
+        if self.tree.views().next().is_none() {
```

Comment:
```suggestion
        if self.tree.views().is_empty() {
```

Reply:
Since `self.tree.views()` returns an `impl Iterator`, this doesn't work:

```
error[E0599]: no method named `is_empty` found for opaque type `impl Iterator` in the current scope
   --> helix-view/src/editor.rs:453:30
    |
453 |         if self.tree.views().is_empty() {
    |                              ^^^^^^^^ method not found in `impl Iterator`

For more information about this error, try `rustc --explain E0599`.
error: could not compile `helix-view` due to previous error
```

---

I will add a comment, though.